snorkel/udf.py

```diff
@@ -1,77 +1,46 @@
-        self.out_queue    = out_queue
-        self.apply_kwargs = {}
-
-        # Each UDF starts its own Engine
-        # See http://docs.sqlalchemy.org/en/latest/core/pooling.html#using-connection-pools-with-multiprocessing
-        SnorkelSession = new_sessionmaker()
-        self.session   = SnorkelSession()
-
-    def run(self):
-        """
-        This method is called when the UDF is run as a Process in a multiprocess setting
-        The basic routine is: get from JoinableQueue, apply, put / add outputs, loop
-        """
-        while True:
-            try:
-                x = self.in_queue.get(True, QUEUE_TIMEOUT)
-                for y in self.apply(x, **self.apply_kwargs):
-
-                    # If an out_queue is provided, add to that, else add to session
-                    if self.out_queue is not None:
-                        self.out_queue.put(y, True, QUEUE_TIMEOUT)
-                    else:
-                        self.session.add(y)
-                self.in_queue.task_done()
-            except Empty:
-                break
-        self.session.commit()
-        self.session.close()
-
-    def apply(self, x):
-        """This function takes in an object, and returns a generator / set / list"""
-        raise NotImplementedError()
-
 
 class UDFRunner(object):
     """Class to run UDFs in parallel using simple queue-based multiprocessing setup"""
-    def __init__(self, udf_class, **udf_kwargs):
-        self.udf_class  = udf_class
-        self.udf_kwargs = udf_kwargs
-        self.udfs       = []
-        self.udf0       = self.udf_class(**self.udf_kwargs) if hasattr(self.udf_class, 'reduce') else None
-
-    def apply(self, xs, parallelism=None, progress_bar=True, **kwargs):
-
-        # Clear everything if UDF has this function
-        if hasattr(self.udf0, 'clear'):
-            self.udf0.clear(xs, **kwargs)
+    def __init__(self, udf_class, **udf_init_kwargs):
+        self.udf_class       = udf_class
+        self.udf_init_kwargs = udf_init_kwargs
+        self.udfs            = []
+        self.udf0            = self.udf_class(**self.udf_init_kwargs) if hasattr(self.udf_class, 'reduce') else None
+
+    def apply(self, xs, clear=True, parallelism=None, progress_bar=True, **kwargs):
+
+        # Clear everything downstream of this UDF if requested
+        if clear:
+            # We make a copy of kwargs because clear() has a catchall for kwargs and UDF.apply() does not
+            clear_kwargs = kwargs.copy()
+            clear_kwargs['clear'] = clear
+
+            SnorkelSession = new_sessionmaker()
+            session = SnorkelSession()
+            self.clear(session, **kwargs)
+            session.commit()
+            session.close()
 
         # Execute the UDF
         if parallelism is None or parallelism < 2:
             self.apply_st(xs, progress_bar=progress_bar, **kwargs)
         else:
             self.apply_mt(xs, parallelism, **kwargs)
 
+    def clear(self, session, **kwargs):
```

Comment:
Makes sense to put this here, circumvents the annoyances that I was facing putting in `UDF`, nice